ganttproject/src/main/java/net/sourceforge/ganttproject/chart/UnboundedViewportTaskSceneBuilder.kt

```diff
@@ -0,0 +1,184 @@
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
+*/
+package net.sourceforge.ganttproject.chart
+
+import biz.ganttproject.core.chart.canvas.Canvas
+import biz.ganttproject.core.chart.render.ShapeConstants
+import biz.ganttproject.core.chart.render.ShapePaint
+import biz.ganttproject.core.chart.scene.BarChartActivity
+import biz.ganttproject.core.chart.scene.BarChartConnector
+import biz.ganttproject.core.chart.scene.gantt.Connector
+import biz.ganttproject.core.chart.scene.gantt.DependencySceneBuilder
+import biz.ganttproject.core.chart.scene.gantt.TaskActivitySceneBuilder
+import biz.ganttproject.core.chart.scene.gantt.TaskLabelSceneBuilder
+import biz.ganttproject.core.option.DefaultEnumerationOption
+import biz.ganttproject.core.time.TimeDuration
+import biz.ganttproject.core.time.TimeDurationImpl
+import biz.ganttproject.core.time.impl.GPTimeUnitStack
+import net.sourceforge.ganttproject.GanttProjectImpl
+import net.sourceforge.ganttproject.task.dependency.TaskDependencyConstraint
+import net.sourceforge.ganttproject.task.dependency.TaskDependency.Hardness
+import java.awt.Color
+import java.awt.Dimension
+import java.awt.Rectangle
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+class UnboundedViewportTaskSceneBuilder(
+  private val input: TaskActivitySceneBuilder.ChartApi,
+  private val logger: ((String) -> Unit)? = null
+) {
+  fun build(tasks: List<Task>, canvas: Canvas = Canvas()): Canvas {
+    val allTasks = tasks.flatMap { it.flatten() }
+    paintTasks(allTasks, canvas)
+    paintDependencies(allTasks, canvas)
+    return canvas
+  }
+
+  private fun paintTasks(tasks: List<Task>, canvas: Canvas) {
+    val taskBarOffset = (input.rowHeight - input.barHeight) / 2
+    val taskActivityScene = TaskActivitySceneBuilder(
+      ActivitySceneTaskApi, input, canvas, TaskLabelSceneBuilderStub, TaskActivitySceneBuilder.Style(taskBarOffset)
+    )
+    tasks.forEachIndexed { row, task ->
+      taskActivityScene.renderActivities(row, listOf(task.activity), input.bottomUnitOffsets)
+      val nextLineY = row * input.rowHeight
+      val nextLine = canvas.createLine(0, nextLineY, input.viewportWidth, nextLineY)
+      nextLine.foregroundColor = Color.GRAY
+    }
+  }
+
+  private fun paintDependencies(tasks: List<Task>, canvas: Canvas) {
+    val dependencyRenderer = DependencySceneBuilder(
+      canvas, canvas.newLayer(), DependencySceneTaskApi(tasks, logger), { input.barHeight }
+    )
+    dependencyRenderer.build()
+  }
+
+  abstract class Task {
+    abstract val start: Date
+    abstract val end: Date
+    abstract val isMilestone: Boolean
+    abstract val subtasks: List<Task>
+    abstract val dependants: List<Dependant>
+
+    val activity = Activity()
+
+    fun flatten(result: MutableList<Task> = mutableListOf()): List<Task> {
```

Comment:
this is basically pre-order DFS? There's `TaskContainmentHierarchyFacade#getTasksInDocumentOrder` which I think should be used.